ion/timestamp.go

```diff
@@ -0,0 +1,303 @@
+	date := time.Date(ts[0], time.Month(ts[1]), ts[2], ts[3], ts[4], ts[5], nsecs, time.UTC)
+	// time.Date converts 2000-01-32 input to 2000-02-01
+	if ts[0] != date.Year() || time.Month(ts[1]) != date.Month() || ts[2] != date.Day() {
+		return emptyTimestamp(), fmt.Errorf("ion: invalid timestamp")
+	}
+
+	if precision <= Day {
+		return NewDateTimestamp(date, precision), nil
+	}
+
+	if overflow {
+		date = date.Add(time.Second)
+	}
+
+	if offset == 0 {
+		if sign == -1 {
+			// Negative zero timezone offset is Unspecified
+			return NewTimestampWithFractionalSeconds(date, precision, Unspecified, fractionPrecision), nil
+		}
+		// Positive zero timezone offset is UTC
+		return NewTimestampWithFractionalSeconds(date, precision, UTC, fractionPrecision), nil
+	}
+
+	date = date.In(time.FixedZone("fixed", int(offset)*60))
+
+	// Non-zero offset is Local
+	return NewTimestampWithFractionalSeconds(date, precision, Local, fractionPrecision), nil
+}
+
+// Format returns a formatted Timestamp string.
+func (ts *Timestamp) Format() string {
+	layout := ts.precision.Layout(ts.kind, ts.numFractionalSeconds)
+	format := ts.dateTime.Format(layout)
+
+	// The above time.Format() does not produce the format we want in some scenarios.
+	// So we may need to make some adjustments.
+
+	// Add back removed trailing zeros from fractional seconds (ie. ".000")
+	if ts.precision >= Nanosecond && ts.dateTime.Nanosecond() == 0 && ts.numFractionalSeconds > 0 {
+		// Find the position of 'T'
+		tIndex := strings.Index(format, "T")
+		if tIndex == -1 {
+			tIndex = strings.Index(format, "t")
+			if tIndex == -1 {
+				return format
+			}
+		}
+
+		index := strings.LastIndex(format, "Z")
+		if index == -1 || index < tIndex {
+			index = strings.LastIndex(format, "+")
+			if index == -1 || index < tIndex {
+				index = strings.LastIndex(format, "-")
+			}
+		}
+
+		// This position better be right of 'T'
+		if index != -1 && tIndex < index {
+			zeros := "."
+			for i := uint8(0); i < ts.numFractionalSeconds; i++ {
+				zeros += "0"
+			}
+
+			format = format[0:index] + zeros + format[index:]
+		}
+	}
+
+	// A timestamp with time precision (ie. Minute/Second/Nanosecond) and Unspecified timezone
+	// should have a "-00:00" offset but time.Format() is returning a "+00:00" offset.
+	if ts.precision >= Minute && ts.kind == Unspecified {
+		index := strings.LastIndex(format, "+00:00")
+		if index != -1 {
+			format = format[0:index] + "-00:00"
+		}
+	}
+
+	return format
+}
+
+// Equal figures out if two timestamps are equal for each component.
+func (ts *Timestamp) Equal(ts1 Timestamp) bool {
+	_, offset := ts.dateTime.Zone()
+	_, offset1 := ts1.dateTime.Zone()
+
+	return ts.dateTime.Equal(ts1.dateTime) &&
+		offset == offset1 &&
+		ts.precision == ts1.precision &&
+		ts.kind == ts1.kind &&
+		ts.numFractionalSeconds == ts1.numFractionalSeconds
+}
+
+// SetLocation sets the location for the internal time object.
+func (ts *Timestamp) SetLocation(loc *time.Location) {
```

Comment:
I quite like that `time.Time` values are immutable. Would it make sense to follow that for `Timestamp` as well, and change this to `func (ts Timestamp) In(loc *time.Location) Timestamp`?

Reply:
I changed the TImestamp functions to use value receivers. 
Out of curiosity, what is your rationale for preferring immutable value receivers over pointer receivers?

Reply:
In general, I prefer immutable values for things like this because they're safe to use as constants, reference from different goroutines without synchronization, etc. That's sketchier if I have to worry about someone else changing the timezone out from underneath me. Go follows this pattern a reasonable amount, e.g. with strings and time.Time.

As far as value-type receiver vs pointer-type, it doesn't make a huge difference in practice, but value-type receivers are a signal that the value is immutable and meant to be passed around by value, ala `time.Time`. (This is a pattern I didn't follow with `Decimal`, although in retrospect I arguably should have. I started off basing that one on `*big.Int`, which allows mutation so you can reuse the underlying storage instead of making copies of large byte arrays. I then switched to an immutable style because all the decimals I was working with were small-number-of-digit amounts of money where the copying wasn't an issue, but didn't switch the receivers. :))

Reply:
Makes sense. Thanks for the explanation.